tsfc/coffee_mode.py

```diff
@@ -0,0 +1,294 @@
+            connection.append(atomic_index[atomic])
+        connections.append(tuple(connection))
+
+    if len(atomic_index) == 0:
+        return ((), ())
+    if len(atomic_index) == 1:
+        return ((next(iterkeys(atomic_index)), ), ())
+
+    # set up the ILP
+    import pulp as ilp
+    ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
+    ilp_var = ilp.LpVariable.dicts('node', range(len(atomic_index)), 0, 1, ilp.LpBinary)
+
+    # Objective function
+    # Minimise number of factors to pull. If same number, favour factor with larger extent
+    big = 1e20  # some arbitrary big number
+    ilp_prob += ilp.lpSum(ilp_var[index] * (big - index_extent(atomic, argument_indices))
+                          for atomic, index in iteritems(atomic_index))
+
+    # constraints
+    for connection in connections:
+        ilp_prob += ilp.lpSum(ilp_var[index] for index in connection) >= 1
+
+    ilp_prob.solve()
+    if ilp_prob.status != 1:
+        raise RuntimeError("Something bad happened during ILP")
+
+    def optimal(atomic):
+        return ilp_var[atomic_index[atomic]].value() == 1
+
+    optimal_atomics = filter(optimal, iterkeys(atomic_index))
+    return tuple(optimal_atomics)
+
+
+def factorise_atomics(monomial_sum, optimal_atomics, argument_indices):
+    """Group and factorise monomials using a list of atomics as common
+    subexpressions. Create new monomials for each group and optimise them recursively.
+
+    :arg monomial_sum: A :class:`MonomialSum` object, or a iterable collection
+                       of monomials, all of the monomials should have the
+                       same sum indices
+    :arg optimal_atomics: list of tuples of atomics to be used as common subexpression
+    :arg argument_indices: tuple of argument indices
+
+    :returns: a factorised :class:`MonomialSum` object, or the original object
+    if no changes are made
+    """
+    if not optimal_atomics or len(monomial_sum) < 2:
+        # Nothing to do
+        if isinstance(monomial_sum, MonomialSum):
+            return monomial_sum
+        else:
+            # input is list of monomials
+            new_monomial_sum = MonomialSum()
+            for m in monomial_sum:
+                new_monomial_sum.add(*m)
+            return new_monomial_sum
+    new_monomial_sum = MonomialSum()
+    # Group monomials with respect to each optimal atomic
+    factor_group = OrderedDict()
+    for monomial in monomial_sum:
+        for oa in optimal_atomics:
+            if oa in monomial.atomics:
+                # Add monomial to the list of corresponding optimal atomic
+                factor_group.setdefault(oa, []).append(monomial)
+                break
+        else:
+            # Add monomials that do no have argument factors to new MonomialSum
+            new_monomial_sum.add(*monomial)
+    # We should not drop monomials
+    assert sum(map(len, itervalues(factor_group))) + len(new_monomial_sum) == len(monomial_sum)
+
```

Comment:
```python
sum_indices = next(iter(monomial_sum)).sum_indices
```

Should probably assert that all the provided monomials have the same sum_indices.